File: android/CartFollow/app/src/main/java/com/cartfollow/tracker/CartProtocol.kt
package com.cartfollow.tracker

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Matches cart_protocol.py / ESP32 usb_data_t framing.
 * Phone sends steering only; Pi LiDAR supplies throttle.
 * Set PERSON_DETECTED_FLAG in button_states when a person is in frame.
 */
object CartProtocol {
    const val START_BYTE: Byte = 0xAA.toByte()
    const val END_BYTE: Byte = 0xBB.toByte()
    const val PERSON_DETECTED_FLAG: Int = 0x80
    const val CALIBRATING_FLAG: Int = 0x40
    const val CALIB_OK_FLAG: Int = 0x20
    const val PI_PORT: Int = 9747

    const val MAX_STEERING = 512
    /** Absolute floor; real deadzone is also a fraction of frame half-width. */
    const val DEADZONE_X = 12
    /** Ignore centering error within this fraction of half-frame (~8% of width). */
    const val DEADZONE_X_FRAC = 0.08f

    // Backup distance control (used only when the Pi LiDAR has no valid reading).
    // Torso height as a fraction of frame height at the desired follow distance.
    // Bigger torso = closer. Tune TARGET_BODY_FRAC for your mount + follow distance.
    const val TARGET_BODY_FRAC = 0.33f
    const val BODY_DEADZONE_FRAC = 0.03f
    const val MAX_BACKUP_THROTTLE = 320 // closer to LiDAR authority when fused
    const val MIN_THROTTLE = 40

    fun buildPacket(
        steering: Int,
        personDetected: Boolean,
        throttle: Int = 0,
        calibrating: Boolean = false,
        calibOk: Boolean = false,
    ): ByteArray {
        val emergencyStop: Byte = 0
        val cruiseControl: Byte = 0
        val cruiseSpeed: Short = 0
        val followMeMode: Byte = 1
        val turboMode: Byte = 0
        val batteryLevel: Byte = 0
        val buttonStates = (
            (if (personDetected) PERSON_DETECTED_FLAG else 0) or
            (if (calibrating) CALIBRATING_FLAG else 0) or
            (if (calibOk) CALIB_OK_FLAG else 0)
        ).toByte()

        val body = ByteBuffer.allocate(12).order(ByteOrder.LITTLE_ENDIAN)
        body.putShort(throttle.toShort())
        body.putShort(steering.toShort())
        body.put(emergencyStop)
        body.put(cruiseControl)
        body.putShort(cruiseSpeed)
        body.put(followMeMode)
        body.put(turboMode)
        body.put(batteryLevel)
        body.put(buttonStates)

        val payload = body.array()
        var checksum = 0
        for (b in payload) {
            checksum = (checksum + (b.toInt() and 0xFF)) and 0xFF
        }

        return byteArrayOf(START_BYTE) + payload + checksum.toByte() + END_BYTE
    }

    fun steeringFromCenter(personCenterX: Float, frameWidth: Int, targetCenterX: Float? = null): Int {
        if (frameWidth <= 0) return 0
        val center = targetCenterX ?: (frameWidth / 2f)
        // Person left of center → positive steer → Pi left turn (keep user in frame
        // when walking the same direction as the cart). Facing the cart mirrors
        // left/right in the world, which is expected.
        val errorX = center - personCenterX
        val deadzone = kotlin.math.max(DEADZONE_X.toFloat(), center * DEADZONE_X_FRAC)
        if (kotlin.math.abs(errorX) <= deadzone) return 0
        val ratio = errorX / center
        var steering = (ratio * MAX_STEERING).toInt()
        steering = steering.coerceIn(-MAX_STEERING, MAX_STEERING)
        return steering
    }

    fun throttleFromBodyFraction(bodyFrac: Float, targetBodyFrac: Float = TARGET_BODY_FRAC): Int {
        if (bodyFrac <= 0f) return 0
        val error = targetBodyFrac - bodyFrac
        if (kotlin.math.abs(error) <= BODY_DEADZONE_FRAC) return 0
        val ratio = error / targetBodyFrac
        var throttle = (ratio * MAX_BACKUP_THROTTLE).toInt()
        if (throttle != 0 && kotlin.math.abs(throttle) < MIN_THROTTLE) {
            // Soft floor: tiny errors stay 0 so fusion doesn't inject creep
            if (kotlin.math.abs(throttle) < MIN_THROTTLE / 2) return 0
            throttle = if (throttle > 0) MIN_THROTTLE else -MIN_THROTTLE
        }
        return throttle.coerceIn(-MAX_BACKUP_THROTTLE, MAX_BACKUP_THROTTLE)
    }
}

File: cart_protocol.py
"""Binary follow-me packet protocol shared by the Android phone and the Pi."""

from __future__ import annotations

import struct
from typing import List, Optional, Tuple

START_BYTE = 0xAA
END_BYTE = 0xBB
PAYLOAD_SIZE = 13

# struct usb_data_t — little-endian, packed (matches hoverboard_minimal / ESP32)
_PAYLOAD_FMT = "<hhBBhBBBB"


def checksum(payload: bytes) -> int:
    return sum(payload[:12]) & 0xFF


def build_packet(
    throttle: int = 0,
    steering: int = 0,
    emergency_stop: int = 0,
    cruise_control: int = 0,
    cruise_speed: int = 0,
    follow_me_mode: int = 1,
    turbo_mode: int = 0,
    battery_level: int = 0,
    button_states: int = 0,
) -> bytes:
    body = struct.pack(
        _PAYLOAD_FMT,
        int(throttle),
        int(steering),
        int(emergency_stop) & 0xFF,
        int(cruise_control) & 0xFF,
        int(cruise_speed),
        int(follow_me_mode) & 0xFF,
        int(turbo_mode) & 0xFF,
        int(battery_level) & 0xFF,
        int(button_states) & 0xFF,
    )
    body += bytes([checksum(body)])
    return bytes([START_BYTE]) + body + bytes([END_BYTE])


def parse_packet(data: bytes) -> Optional[dict]:
    """Validate framing and return decoded fields, or None if invalid."""
    if len(data) < 15 or data[0] != START_BYTE or data[-1] != END_BYTE:
        return None

    payload = data[1:-1]
    if len(payload) != PAYLOAD_SIZE:
        return None

    fields = struct.unpack(_PAYLOAD_FMT, payload[:12])
    if (sum(payload[:12]) & 0xFF) != payload[12]:
        return None

    return {
        "throttle": fields[0],
        "steering": fields[1],
        "emergency_stop": fields[2],
        "cruise_control": fields[3],
        "cruise_speed": fields[4],
        "follow_me_mode": fields[5],
        "turbo_mode": fields[6],
        "battery_level": fields[7],
        "button_states": fields[8],
        "checksum": payload[12],
    }


def extract_packets(buffer: bytearray) -> Tuple[List[bytes], bytearray]:
    """Pull complete framed packets from a growing byte buffer."""
    packets = []
    while True:
        start = buffer.find(bytes([START_BYTE]))
        if start == -1:
            return packets, bytearray()
        if start > 0:
            del buffer[:start]
        end = buffer.find(bytes([END_BYTE]), 1)
        if end == -1:
            return packets, buffer
        packet = bytes(buffer[: end + 1])
        del buffer[: end + 1]
        packets.append(packet)
